crates/rust-analyzer/src/config.rs

```diff
@@ -1565,8 +1594,13 @@ impl Config {
                             .ok()
                             .map(Into::into)
                     }
-                    ManifestOrProjectJson::ProjectJson(it) => {
-                        Some(ProjectJson::new(None, &self.root_path, it.clone()).into())
+                    ManifestOrProjectJson::ProjectJson { data, buildfile } => {
+                        let root_path = match buildfile {
+                            Some(buildfile) => AbsPathBuf::try_from(buildfile.clone())
+                                .expect("unable to get AbsPathBuf"),
+                            None => self.root_path.clone(),
+                        };
+                        Some(ProjectJson::new(None, &root_path, data.clone()).into())
```

Comment:
I wrote this in a haze, but I don't think it should be necessary to clone here.

Reply:
this is less gross now, but a conversion from an `Utf8PathBuf` to an `AbsPathBuf is still unfortunately necessary. Fixing that would be a larger refactor, but it's not one I'm opposed to making.